python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+        """
+        return self.output_map.get(task_id, self.checkpoint_map.get(task_id))
+
+    def pop_frontier_to_run(self) -> Optional[StepID]:
+        """Pop one task to run from the frontier queue."""
+        try:
+            t = self.frontier_to_run.popleft()
+            self.frontier_to_run_set.remove(t)
+            return t
+        except IndexError:
+            return None
+
+    def append_frontier_to_run(self, task_id: StepID) -> None:
+        """Insert one task to the frontier queue."""
+        if (
+            task_id not in self.frontier_to_run_set
+            and task_id not in self.running_frontier_set
+        ):
+            self.frontier_to_run.append(task_id)
+            self.frontier_to_run_set.add(task_id)
+
+    def add_dependencies(self, task_id: StepID, in_dependencies: List[StepID]) -> None:
+        """Add dependencies between a task and it input dependencies."""
+        self.in_dependency[task_id] = in_dependencies
+        for in_task_id in in_dependencies:
+            self.out_dependency[in_task_id].append(task_id)
+
+    def pop_running_frontier(self, fut: asyncio.Future) -> WorkflowRef:
+        """Pop a task from the running frontier."""
+        ref = self.running_frontier.pop(fut)
+        self.running_frontier_set.remove(ref.task_id)
+        return ref
+
+    def insert_running_frontier(self, fut: asyncio.Future, ref: WorkflowRef) -> None:
+        """Insert a task to the running frontier."""
+        self.running_frontier[fut] = ref
+        self.running_frontier_set.add(ref.task_id)
+
+    def append_continuation(
+        self, task_id: StepID, continuation_task_id: StepID
+    ) -> None:
+        """Append continuation to a task."""
+        continuation_root = self.continuation_root.get(task_id, task_id)
+        self.prev_continuation[continuation_task_id] = task_id
+        self.next_continuation[task_id] = continuation_task_id
+        self.continuation_root[continuation_task_id] = continuation_root
+        self.latest_continuation[continuation_root] = continuation_task_id
+        # migrate callbacks
+        if continuation_task_id in self.done_task_callbacks:
+            self.done_task_callbacks[
+                self.continuation_root[continuation_task_id]
+            ].extend(self.done_task_callbacks.pop(continuation_task_id))
+
+    def merge_state(self, state: "WorkflowExecutionState") -> None:
+        """Merge with another execution state."""
+        self.in_dependency.update(state.in_dependency)
+        self.out_dependency.update(state.out_dependency)
+        self.task_input_args.update(state.task_input_args)
+        self.task_func_body.update(state.task_func_body)
+        self.task_options.update(state.task_options)
+        self.task_context.update(state.task_context)
+        self.output_map.update(state.output_map)
+        self.checkpoint_map.update(state.checkpoint_map)
+
+    def construct_runtime_deps(self, task_id: StepID) -> None:
```

Comment:
```suggestion
    def construct_upstream_tasks(self, task_id: StepID) -> None:
```

(runtime_deps could also mean something like packages)